keel-sql/src/test/kotlin/com/netflix/spinnaker/keel/sql/MigratorErrorTests.kt

```diff
@@ -0,0 +1,128 @@
+import strikt.assertions.isFailure
+import strikt.assertions.isSuccess
+import java.time.Clock.fixed
+import java.time.Duration.ofMinutes
+import java.time.Instant.now
+import java.time.ZoneOffset.UTC
+
+/**
+ * Tests that exceptions thrown by migrators are handled gracefully when reading from the database.
+ */
+class MigratorErrorTests {
+  private val jooq = testDatabase.context
+  private val retryProperties = RetryProperties(5, 0)
+  private val objectMapper = configuredTestObjectMapper()
+  private val sqlRetry = SqlRetry(SqlRetryProperties(retryProperties, retryProperties))
+  private val clock = fixed(now(), UTC)
+
+  private val kindV1 = kind<DummyResourceSpec>(parseKind("test/whatever@v1"))
+  private val kindV2 = kind<DummyResourceSpec>(parseKind("test/whatever@v2"))
+
+  private val multiVersionResourceSpecIdentifier = ResourceSpecIdentifier(kindV1, kindV2)
+
+  private val bedShittingSpecMigrator =
+    object : SpecMigrator<DummyResourceSpec, DummyResourceSpec> {
+      override val input = kindV1
+      override val output = kindV2
+
+      override fun migrate(spec: DummyResourceSpec): DummyResourceSpec {
+        throw RuntimeException("💩🛏")
+      }
+    }
+
+  private val resourceRepository = SqlResourceRepository(
+    jooq = jooq,
+    clock = clock,
+    resourceSpecIdentifier = multiVersionResourceSpecIdentifier,
+    specMigrators = listOf(bedShittingSpecMigrator),
+    objectMapper = objectMapper,
+    sqlRetry = sqlRetry
+  )
+
+  private val deliveryConfigRepository = SqlDeliveryConfigRepository(
+    jooq = jooq,
+    clock = clock,
+    resourceSpecIdentifier = multiVersionResourceSpecIdentifier,
+    objectMapper = objectMapper,
+    sqlRetry = sqlRetry,
+    artifactSuppliers = defaultArtifactSuppliers(),
+    specMigrators = listOf(bedShittingSpecMigrator)
+  )
+
+  val deliveryConfig = DeliveryConfig(
+    name = "fnord-manifest",
+    application = "fnord",
+    serviceAccount = "keel@spinnaker",
+    environments = setOf(
+      Environment(
+        name = "test",
+        resources = setOf(
+          resource(kind = kindV1.kind),
+          resource(kind = kindV2.kind)
+        )
+      )
+    )
+  )
+
+  @BeforeEach
+  fun persistData() {
+    deliveryConfig.environments.first().resources.forEach {
+      resourceRepository.store(it)
+    }
+    deliveryConfigRepository.store(deliveryConfig)
+  }
+
+  @Test
+  fun `the failing delivery config does not block the environment check cycle`() {
+    // The first time our un-readable delivery config will be due for a check, so the method should fail
+    expectCatching {
+      deliveryConfigRepository.itemsDueForCheck(ofMinutes(1), 10)
+    }
+      .isFailure()
+      .isA<RuntimeException>()
+
+    // Subsequently we should not try to re-check the same delivery config (within the time window)
+    expectCatching {
+      deliveryConfigRepository.itemsDueForCheck(ofMinutes(1), 10)
+    }
+      .isSuccess()
+      .isEmpty()
+  }
```

Comment:
Here's the test that confirms the behavior works as desired (and which failed before moving the load outside the tx).